middleware/src/middleware.go

```diff
@@ -556,6 +563,154 @@ func (middleware *Middleware) RebootBase() rpcmessages.ErrorResponse {
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "Could not start the update process. Please see the Middleware log for more detail.",
+			Code:    rpcmessages.ErrorMenderUpdateInstallFailed,
+		}
+	}
+	defer stdout.Close()
+	stdoutScanner := bufio.NewScanner(stdout)
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		log.Printf("Could not get the StderrPipe to read command progress from: %s", err.Error())
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "Could not start the update process. Please see the Middleware log for more detail.",
+			Code:    rpcmessages.ErrorMenderUpdateInstallFailed,
+		}
+	}
+	defer stderr.Close()
+	stderrScanner := bufio.NewScanner(stderr)
+
+	err = cmd.Start()
+	if err != nil {
+		log.Printf("Could not run the Base update command: %s", err.Error())
+		return rpcmessages.ErrorResponse{
+			Success: false,
+			Message: "Could not start the update process. Please see the Middleware log for more detail.",
+			Code:    rpcmessages.ErrorMenderUpdateInstallFailed,
+		}
+	}
+
+	errOutLines := make([]string, 0)
+	// This goroutine uses the bufio.Scanner to .Scan() `stderr` lines.
+	// This is done in a goroutine, since .Scan() blocks when there is no input available.
+	// Every line read is appended to errOutLines, a string slice with stderr lines.
+	// The goroutine exits once EOF for `stderr` is reached.
+	// Once `stdout` reaches EOF the `stderr` pipe is closed (see below).
+	go func() {
+		for {
+			hasReadSomething := stderrScanner.Scan()
+			if hasReadSomething {
+				// Lines written to stderr are captured in errOutLines and processed if os.Wait returns an error
+				lineErr := stderrScanner.Text()
+				errOutLines = append(errOutLines, strings.TrimSuffix(lineErr, "\n"))
+			} else {
+				if stderrScanner.Err() != nil {
+					log.Printf("GetBaseUpdateProgress: Could not read from stderr scanner: %s", stderrScanner.Err())
+					stderr.Close()
+					return
+				}
+				// When scanner.Scan() returns `false` and scanner.Err() is `nil` then EOF of `stderr` is reached. The goroutine exits.
+				return
+			}
+		}
+	}()
+
+	for {
+		hasReadSomething := stdoutScanner.Scan()
+		if hasReadSomething {
+			lineOut := stdoutScanner.Text()
+			log.Println(lineOut)
+			containsProgressUpdateInfo, percentage, downloadedKiB := parseBaseUpdateStdout(lineOut)
+			if containsProgressUpdateInfo {
+				middleware.baseUpdateProgress.ProgressPercentage = percentage
+				middleware.baseUpdateProgress.ProgressDownloadedKiB = downloadedKiB
+				if percentage < 98 {
+					middleware.setBaseUpdateStateAndNotify(rpcmessages.UpdateDownloading)
+				} else {
+					// switch State from `UpdateDownloading` to `UpdateApplying`
+					// This is done at 98% or above since the mender-install script does
+					// only log 100% after applying the update.
+					middleware.setBaseUpdateStateAndNotify(rpcmessages.UpdateApplying)
+				}
+			}
```

Comment:
How does this for loop break in the normal case?

Reply:
Once EOF is read from stdout. https://github.com/digitalbitbox/bitbox-base/pull/228/files/79f263847831dd1862a21fd202ddec22be28544b#diff-bb48a3164edaae7002e0a4949d26fd9dR683-R685

Will add a comment to the for-loop clarifying that.